src/driver/linux_resource/syscall_aarch64.c

```diff
@@ -142,53 +461,122 @@ find_syscall_table_via_vbar(void)
 {
   ci_uint8 *el_svc = find_el_svc_entry();
   ci_uint32 insn;
+#if LINUX_VERSION_CODE >= KERNEL_VERSION(5, 4, 0)
+  ci_uint8 *do_el0_svc = NULL;
+  ci_uint8 *sys_call_table = NULL;
+#endif
 
   if (el_svc == NULL)
     return NULL;
 
-  if (aarch64_insn_read_sym(el_svc, &insn)) {
+  /* syscall entry code was rewritten to C in 5.4. So we need to use another
+   * approach for sys_call_table look up.  */
+#if LINUX_VERSION_CODE >= KERNEL_VERSION(5, 4, 0)
+
+  /*
+   * el0_svc() calls do_el0_svc() after some code. It is the 3d 'bl' instruction:
+   * <el0_svc>:
+   * d503233f        paciasp
+   * a9be7bfd        stp     x29, x30, [sp, #-32]!
+   * 910003fd        mov     x29, sp
+   * f9000bf3        str     x19, [sp, #16]
+   * aa0003f3        mov     x19, x0
+   * 97d930ee        bl      ffffffc0081b9360 <trace_hardirqs_off_finish>
+   * 97d2ae0d        bl      ffffffc0080187e0 <cortex_a76_erratum_1463225_svc_handler>
+   * aa1303e0        mov     x0, x19
+   * 97d2ebbc        bl      ffffffc008027ea4 <do_el0_svc>
+   */
+  el_svc = find_next_bl(el_svc);
+  el_svc += AARCH64_INSN_SIZE;
+
+  el_svc = find_next_bl(el_svc);
+  el_svc += AARCH64_INSN_SIZE;
+
+  el_svc = find_next_bl(el_svc);
+  CI_AARCH64_INSN_READ(el_svc, insn);
+
+  /* Jumping to do_el0_svc(). */
+  do_el0_svc = el_svc + ci_aarch64_get_branch_offset(insn);
+  TRAMP_DEBUG("%s:%d: 'bl do_el0_svc' = %px (%x)", __FUNCTION__, __LINE__,
+              el_svc, insn);
+  TRAMP_DEBUG("%s:%d: do_el0_svc = %px", __FUNCTION__, __LINE__, do_el0_svc);
+
+  /* Next (and the last) we must find the function call where sys_call_table
```

Comment:
This PR is a decent dive into low-level machinery, and I'm afraid I can't allocate more CPU cycles to review it. I trust that it works correctly!

Generally, what symptoms do you observe if something in the new code goes off the rails? Is it an instant kernel panic?

Reply:
No, kernel won't panic, just stop module loading.

Example:
```bash
$ sudo insmod ./aarch64_linux-5.10.110-v8-ol+/driver/linux/sfc_resource.ko 
insmod: ERROR: could not insert module ./aarch64_linux-5.10.110-v8-ol+/driver/linux/sfc_resource.ko: Unknown symbol in module
```
dmesg:
```
[190409.448360] [sfc efrm] init_sfc_resource: RESOURCE driver starting
[190409.454665] [sfc efrm] find_el_sync: el0_sync entry is not a branch
[190409.461037] [sfc efrm] init_sfc_resource: ERROR: failed to find syscall table
```

Reply:
Thanks, it's good we don't make the host unusable after a problematic run.